File: app/src/main/java/com/example/timmytruong/wordsearch_19/MainActivity.java
package com.example.timmytruong.wordsearch_19;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.GridView;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Set;

public class MainActivity extends AppCompatActivity implements MainController.MainInterface {

    /**
     * MainController Declaration
     */
    MainController mainController;

    /**
     * UI Declarations
     */
    GridView gridView;
    FrameLayout gridFrame;
    TextView scoreTV;
    Button resetBTN, plusBTN;
    TableLayout wordTableLayout;

    /**
     * Context declaration
     */
    public static Context context;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        /**
         * Static context that can be used in other classes
         */
        MainActivity.context = getApplicationContext();

        /**
         * UI Definitions
         */
        scoreTV = findViewById(R.id.score);
        resetBTN = findViewById(R.id.resetBTN);
        plusBTN = findViewById(R.id.plusBTN);
        gridView = findViewById(R.id.gridView);
        gridFrame = findViewById(R.id.gridFrame);
        wordTableLayout = findViewById(R.id.wordTableLayout);

        /**
         * Controller Object and Method Call
         */
        mainController = new MainController(this, gridView, gridFrame, resetBTN, plusBTN, scoreTV, wordTableLayout);
        mainController.setGrid();
    }

    /**
     * DESCRIPTION: Handles results after the edit activity is closed
     * @param requestCode: ??
     * @param resultCode: What the request code was when sent
     * @param data: The data that was sent back.
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == 0) {
            ArrayList<String> words = data.getStringArrayListExtra("words");
            mainController.reset(words);
        } else {
            mainController.reset();
        }
    }

    /**
     * DESCRIPTION: Setting the grid view to the created array of letters.
     * @param letterAdapter: Takes care of setting the array, setting tags, and every letter
     * @param gridView: Holds the letters in the form of a letter adapter object
     */
    @Override
    public void setLetters(LetterAdapter letterAdapter, GridView gridView) {
        gridView.setAdapter(letterAdapter);
    }

    /**
     * DESCRIPTION: Setting the onTouchListener (Controls the paint)
     * @param onTouchListener: The on touch listener detects movement inputted from the user and paints accordingly (also checks for correct answer)
     * @param gridView: Holds the views
     */
    @Override
    public void setOnTouchListener(View.OnTouchListener onTouchListener, GridView gridView) {
        gridView.setOnTouchListener(onTouchListener);
    }

    /**
     * DESCRIPTION: Setting the Reset button's on click listener
     * @param onClickListener: Controls the function of the button, Resetting the board, score, and rearranging the letters
     * @param resetBTN: Reset BTN View
     */
    @Override
    public void setResetClickListener(View.OnClickListener onClickListener, Button resetBTN) {
        resetBTN.setOnClickListener(onClickListener);
    }

    /**
     * DESCRIPTION: Sets the score text view
     * @param score: Every time the score is updated, it is passed in to be set
     * @param scoreTV: Score textView UI
     */
    @Override
    public void setScoreTextView(int score, int total, TextView scoreTV) {
        scoreTV.setText(Integer.toString(score) + "/" + Integer.toString(total));
    }

    /**
     * DESCRIPTION: Removes the paint search view based on what is needed to be removed
     * @param startViewCount: The starting point of what views need to be removed
     * @param endViewCount: The ending point of what views need to be removed
     * @param gridFrame: The FrameLayout which holds all these views
     */
    @Override
    public void removeSearchView(int startViewCount, int endViewCount, FrameLayout gridFrame) {
        gridFrame.removeViews(startViewCount, endViewCount - startViewCount);
    }

    /**
     * DESCRIPTION: Strikes out the text whenever a word is successfully found
     * @param word: The word that is found
     * @param tableLayout: The  layout in which to find the textview
     */
    @Override
    public void strikeOutWord(String word, TableLayout tableLayout) {

        // Loop through the tablelayout to get the tablerow
        for (int i = 0; i < tableLayout.getChildCount(); i++) {

            // The tableRow
            View rowView = tableLayout.getChildAt(i);

            // If the table row is a tablerow
            if (rowView instanceof TableRow) {

                // The tableRow
                TableRow row = (TableRow) rowView;

                // Loop through the table row to get the textviews
                for (int k = 0; k < 2; k++) {

                    // The Textview
                    View textView = row.getChildAt(k);

                    // If the textview is a textview and if it equals the word found
                    if (textView instanceof TextView && ((TextView) textView).getText().toString().equals(word)) {

                        // The textview
                        TextView text = (TextView) textView;

                        // Striking the textview out
                        text.setPaintFlags(text.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
                    }
                }
            }
        }
    }

    /**
     * DESCRIPTION: Sets the tablelayout with all the words needing to be found
     * @param keySet: The words that need to be found, in a set
     * @param tableLayout: The table layout that holds the words
     */
    @Override
    public void setTableLayout(Set<String> keySet, TableLayout tableLayout) {
        int wordCounter = 0;

        // Initializing the tablerow
        TableRow tableRow = new TableRow(this);

        // Looping through the keyset of words
        for (String word : keySet) {
            // If the wordcounter is divisible by 2, then make a new row
            if (wordCounter % 2 == 0) {

                // Initializing the table row
                tableRow = new TableRow(this);

                // Setting the parameters of the tablerow
                TableRow.LayoutParams rowParams = new TableRow.LayoutParams();
                rowParams.height = TableLayout.LayoutParams.WRAP_CONTENT;
                rowParams.width = TableLayout.LayoutParams.MATCH_PARENT;
                tableRow.setLayoutParams(rowParams);
            }

            // Making a new textview for each word
            TextView wordText = new TextView(this);

            // Setting the parameters for the textview
            TableRow.LayoutParams textParams = new TableRow.LayoutParams();
            textParams.weight = (float) 0.5;
            textParams.height = TableLayout.LayoutParams.WRAP_CONTENT;
            textParams.width = 0;
            wordText.setLayoutParams(textParams);

            // Setting the settings of the textview
            wordText.setTextSize(16);
            wordText.setText(word);
            wordText.setPadding(0, 10, 0, 10);
            wordText.setGravity(Gravity.CENTER);

            // Removing the parents of each view we created to be able to add each to another parent
            if (wordText.getParent() != null) {
                ((ViewGroup) wordText.getParent()).removeView(wordText);
            }
            if (tableRow.getParent() != null) {
                ((ViewGroup) tableRow.getParent()).removeView(tableRow);
            }

            // Adding the textview to the tablerow
            tableRow.addView(wordText);

            // Adding the tablerow to the tablelayout
            tableLayout.addView(tableRow);

            // Incrementing the wordcounter
            wordCounter++;
        }
    }

    /**
     * DESCRIPTION: Removes all table rows from the table layout (WORDBANK)
     * @param tableLayout: holds the table rows and textviews with words
     */
    @Override
    public void clearTableLayout(TableLayout tableLayout) {
        tableLayout.removeAllViews();
    }

    /**
     * DESCRIPTION: Sets the onclicklistener for the plus button (Edit Words)
     * @param wordsAL: words array list to be editted in the other activity
     * @param plusBTN: plus button UI to set the listener
     */
    @Override
    public void setPlusClickListener(final ArrayList<String> wordsAL, Button plusBTN) {
        View.OnClickListener plusClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, EditWordsActivity.class);
                intent.putStringArrayListExtra("words", wordsAL);
                startActivityForResult(intent, 0);
            }
        };
        plusBTN.setOnClickListener(plusClickListener);
    }

    /**
     * DESCRIPTION: Displays a dialogue box whenever the user wins
     */
    @Override
    public void displayWinDialogue() {
        new AlertDialog.Builder(MainActivity.this)
                .setTitle(R.string.win_title)
                .setMessage(R.string.win_message)
                .setPositiveButton(R.string.play_again, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mainController.reset();
                    }
                })
                .setNegativeButton(R.string.cancel, null)
                .setIcon(android.R.drawable.ic_dialog_info)
                .show();
    }
}

File: app/src/main/java/com/example/timmytruong/wordsearch_19/activity/MainActivity.kt
package com.example.timmytruong.wordsearch_19.activity

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.graphics.Paint
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.timmytruong.wordsearch_19.R
import com.example.timmytruong.wordsearch_19.adapters.GridAdapter
import com.example.timmytruong.wordsearch_19.dagger.component.DaggerAppComponent
import com.example.timmytruong.wordsearch_19.fragment.SettingsFragment
import com.example.timmytruong.wordsearch_19.interfaces.GridHandler
import com.example.timmytruong.wordsearch_19.interfaces.InformationBarHandler
import com.example.timmytruong.wordsearch_19.interfaces.SaveHandler
import com.example.timmytruong.wordsearch_19.model.Letter
import com.example.timmytruong.wordsearch_19.model.Word
import com.example.timmytruong.wordsearch_19.utils.constant.AppConstants
import com.example.timmytruong.wordsearch_19.utils.ui.LetterAdapter
import com.example.timmytruong.wordsearch_19.viewmodel.GridViewModel
import com.example.timmytruong.wordsearch_19.viewmodel.InformationBarViewModel
import com.example.timmytruong.wordsearch_19.viewmodel.factory.GridViewModelFactory
import com.example.timmytruong.wordsearch_19.viewmodel.factory.InformationBarViewModelFactory
import kotlinx.android.synthetic.main.activity_main.*
import javax.inject.Inject

class MainActivity : AppCompatActivity()
{
    @Inject lateinit var gridViewModelFactory: GridViewModelFactory

    @Inject lateinit var informationBarViewModelFactory: InformationBarViewModelFactory

    private lateinit var gridViewModel: GridViewModel

    private lateinit var informationBarViewModel: InformationBarViewModel

    private lateinit var gridAdapter: GridAdapter

    private lateinit var settingsFragment: SettingsFragment

    private val gridHandler = object : GridHandler
    {
        override fun strikeOutWord(context: Context, word: String, tableLayout: TableLayout)
        {
            for (eachRow in 0 until tableLayout.childCount)
            {
                val rowView: View = tableLayout.getChildAt(eachRow)

                if (rowView is TableRow)
                {
                    val tableRow: TableRow = rowView

                    for (eachCell in 0 until rowView.childCount)
                    {
                        val textView: TextView = tableRow.getChildAt(eachCell) as TextView

                        if (textView.text.toString() == word)
                        {
                            val text: TextView = textView

                            text.paintFlags = Paint.STRIKE_THRU_TEXT_FLAG
                        }
                    }
                }
            }
        }

        @SuppressLint("ClickableViewAccessibility")
        override fun setOnTouchListener(context: Context,
                                        onTouchListener: View.OnTouchListener,
                                        gridView: GridView)
        {
            gridView.setOnTouchListener(onTouchListener)
        }

        override fun setLetters(context: Context,
                                letterAdapter: LetterAdapter,
                                gridView: GridView)
        {
            gridView.adapter = letterAdapter
        }

        override fun setTableLayout(context: Context,
                                    keySet: ArrayList<Word>,
                                    tableLayout: TableLayout)
        {

            tableLayout.removeAllViews()

            var tableRow = TableRow(context)

            for ((wordCounter, word) in keySet.withIndex())
            {
                if (wordCounter % 2 == 0)
                {
                    tableRow = TableRow(context)

                    val rowParams: TableRow.LayoutParams = TableRow.LayoutParams()
                    rowParams.height = TableRow.LayoutParams.WRAP_CONTENT
                    rowParams.width = TableRow.LayoutParams.MATCH_PARENT
                    tableRow.layoutParams = rowParams
                }

                val wordText = TextView(context)

                val textParams: TableRow.LayoutParams = TableRow.LayoutParams()
                textParams.weight = 0.5.toFloat()
                textParams.height = TableLayout.LayoutParams.WRAP_CONTENT
                textParams.width = 0
                wordText.layoutParams = textParams

                wordText.textSize = 16.toFloat()
                wordText.text = word.word
                wordText.setPadding(0, 10, 0, 10)
                wordText.gravity = Gravity.CENTER

                if (wordText.parent != null)
                {
                    (wordText.parent as ViewGroup).removeView(wordText)
                }
                if (tableRow.parent != null)
                {
                    (tableRow.parent as ViewGroup).removeView(tableRow)
                }

                tableRow.addView(wordText)

                tableLayout.addView(tableRow)

            }
        }

        override fun clearTableLayout(context: Context,
                                      tableLayout: TableLayout)
        {
            tableLayout.removeAllViews()
        }

        override fun removeSearchView(context: Context,
                                      startViewCount: Int,
                                      endViewCount: Int,
                                      gridFrame: FrameLayout)
        {
            gridFrame.removeViews(startViewCount, endViewCount - startViewCount)
        }

        override fun displayWinDialogue(context: Context)
        {
            val builder = AlertDialog.Builder(context)
                    .setTitle(R.string.win_title)
                    .setMessage(R.string.win_message)
                    .setIcon(android.R.drawable.ic_dialog_alert)

            builder.setPositiveButton(R.string.play_again)
            { _, _ ->
                gridAdapter.reset()
                reset()
            }
            builder.setNegativeButton(R.string.cancel) { _, _ -> }
            builder.show()
        }

        override fun getLetterAdapter(context: Context,
                                      letters: ArrayList<Letter>): LetterAdapter
        {
            return LetterAdapter(context = context,
                    letters = letters)
        }
    }

    private var informationBarHandler: InformationBarHandler = object : InformationBarHandler
    {
        override fun setResetClickListener(resetBTN: Button)
        {
            val resetClickListener = View.OnClickListener {
                gridAdapter.reset()

                reset()
            }
            resetBTN.setOnClickListener(resetClickListener)

        }

        override fun setScoreTextView(score: Int, total: Int, scoreView: TextView)
        {
            val text = "$score/$total"
            scoreView.text = text
        }

        override fun setPlusClickListener(context: Context, words: ArrayList<Word>,
                                          plusBTN: Button)
        {
            val plusClickListener = View.OnClickListener {

                val bundle = Bundle()

                bundle.putSerializable(AppConstants.INTENT_EXTRA_WORDS_ARRAY_LIST_KEY, gridViewModel.getWords())

                settingsFragment.arguments = bundle

                this@MainActivity.supportFragmentManager.beginTransaction()
                        .show(settingsFragment)
                        .commit()
            }
            plusBTN.setOnClickListener(plusClickListener)
        }
    }

    private val saveHandler: SaveHandler = object: SaveHandler
    {
        override fun onSaveClicked()
        {
            gridAdapter.reset()

            reset()
        }

    }

    @Suppress("PLUGIN_WARNING")
    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        DaggerAppComponent.create().inject(this)

        gridViewModel = gridViewModelFactory.create(GridViewModel::class.java)

        informationBarViewModel = informationBarViewModelFactory.create(InformationBarViewModel::class.java)

        settingsFragment = SettingsFragment(saveHandler = saveHandler)

        val bundle = Bundle()

        bundle.putSerializable(AppConstants.INTENT_EXTRA_WORDS_ARRAY_LIST_KEY, gridViewModel.getWords())

        settingsFragment.arguments = bundle

        this@MainActivity.supportFragmentManager.beginTransaction()
                .add(main_screen.id, settingsFragment, AppConstants.FRAGMENT_EDIT_WORDS)
                .hide(settingsFragment)
                .commit()

        gridAdapter = GridAdapter(context = this,
                gridHandler = gridHandler,
                gridView = gridView as GridView,
                gridFrameLayout = gridFrame as FrameLayout,
                wordTableLayout = wordTableLayout as TableLayout,
                scoreTextView = score as TextView,
                gridViewModel = gridViewModel,
                informationBarHandler = informationBarHandler,
                informationBarViewModel = informationBarViewModel)

        gridAdapter.setupWordGrid()

        gridAdapter.setupUI()

        informationBarHandler.setScoreTextView(score = informationBarViewModel.getScore(),
                total = informationBarViewModel.getTotal(),
                scoreView = score)

        informationBarHandler.setResetClickListener(resetBTN = resetBTN)

        informationBarHandler.setPlusClickListener(context = this,
                words = gridViewModel.getWords(),
                plusBTN = plusBTN)
    }

    private fun reset()
    {
        informationBarViewModel.setScore(resetScore = true)

        informationBarHandler.setScoreTextView(score = informationBarViewModel.getScore(),
                total = informationBarViewModel.getTotal(),
                scoreView = score)

        gridHandler.removeSearchView(context = this,
                startViewCount = 1,
                endViewCount = gridFrame.childCount,
                gridFrame = gridFrame)
    }
}